libbeat/processors/add_observer_metadata/add_observer_metadata.go

```diff
@@ -0,0 +1,151 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package add_observer_metadata
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+	"github.com/elastic/beats/libbeat/processors/util"
+	"github.com/elastic/go-sysinfo"
+)
+
+func init() {
+	processors.RegisterPlugin("add_observer_metadata", New)
+}
+
+type observerMetadata struct {
+	lastUpdate struct {
+		time.Time
+		sync.Mutex
+	}
+	data    common.MapStrPointer
+	geoData common.MapStr
+	config  Config
+}
+
+const (
+	processorName = "add_observer_metadata"
+)
+
+// New creates a new instance of the add_observer_metadata processor.
+func New(cfg *common.Config) (processors.Processor, error) {
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrapf(err, "fail to unpack the %v configuration", processorName)
+	}
+
+	p := &observerMetadata{
+		config: config,
+		data:   common.NewMapStrPointer(nil),
+	}
+	p.loadData()
+
+	if config.Geo != nil {
+		geoFields, err := util.GeoConfigToMap(*config.Geo)
+		if err != nil {
+			return nil, err
+		}
+
+		p.geoData = common.MapStr{"observer": common.MapStr{"geo": geoFields}}
+	}
+
+	return p, nil
+}
+
+// Run enriches the given event with the observer meta data
+func (p *observerMetadata) Run(event *beat.Event) (*beat.Event, error) {
+	err := p.loadData()
+	if err != nil {
+		return nil, err
+	}
+
+	keyExists, _ := event.Fields.HasKey("observer")
+
+	if p.config.Overwrite || !keyExists {
+		event.Fields.DeepUpdate(p.data.Get().Clone())
```

Comment:
Now reading the code I wonder if in case we have `overwrite: true` if we should "extend" the object or actually completely overwrite it, meaning first delete `observer`. If we don't delete, we have the risk of having mixed data in the observer object.

Reply:
I agree, overwrite should not extend, I'll patch this to delete first.